nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -564,14 +733,351 @@ impl DataStore {
+
+                // If no local storage allocation is required, then simply try
+                // allocating a VMM to this sled.
+                //
+                // Try to INSERT the record. If this is still a valid target,
+                // we'll use it. If it isn't a valid target, we'll shrink the
+                // set of viable sled targets and try again.
+                let rows_inserted = sled_insert_resource_query(
+                    &resource,
+                    &LocalStorageAllocationRequired::No,
+                )
                 .execute_async(&*conn)
                 .await?;
-            if rows_inserted > 0 {
-                return Ok(resource);
+
+                if rows_inserted > 0 {
+                    info!(&log, "reservation succeeded!");
+                    return Ok(resource);
+                }
+                info!(&log, "reservation failed");
+            } else {
+                // If local storage allocation is required, match the requests
+                // with all the zpools of this sled that have available space.
+                // This routine finds all possible configurations that would
+                // satisfy the requests for local storage and tries them all.
+
+                // First, each request for local storage can possibly be
+                // satisfied by a number of zpools on the sled. Find this list
+                // of candidate zpools for each required local storage
+                // allocation.
+
+                #[derive(Clone, Debug, Hash, PartialEq, Eq)]
+                struct CandidateDataset {
+                    rendezvous_local_storage_dataset_id: DatasetUuid,
+                    pool_id: ZpoolUuid,
+                    sled_id: SledUuid,
+                }
+
+                #[derive(Clone, Debug)]
+                struct PossibleAllocationsForRequest<'a> {
+                    request: &'a LocalStorageDisk,
+                    candidate_datasets: HashSet<CandidateDataset>,
+                }
+
+                let zpools_for_sled = self
+                    .zpool_get_for_sled_reservation(&opctx, sled_target)
+                    .await?;
+
+                // If there's an existing local storage allocation on a zpool,
+                // remove that from the list of candidates. Local storage for
+                // the same Instance should not share any zpools.
+                let local_storage_zpools_used: HashSet<ZpoolUuid> =
+                    local_storage_disks
+                        .iter()
+                        .filter_map(|disk| {
+                            disk.local_storage_dataset_allocation.as_ref().map(
+                                |allocation| {
+                                    ZpoolUuid::from_untyped_uuid(
+                                        allocation
+                                            .pool_id()
+                                            .into_untyped_uuid(),
+                                    )
```

Comment:
oh, this is...turning [an `ExternalZpoolUuid`](https://github.com/oxidecomputer/omicron/blob/827bb37a692616810fb3d8fed79166a211be9fe7/nexus/db-model/src/local_storage_dataset_allocation.rs#L75-L77) into a `ZpoolUuid`? out of curiosity, what is the difference between those?

Reply:
thanks, yeah, that makes sense. carry on!